### PR TITLE
Fix re-registering of device after an exit signal in the `ExUnit.CaptureIO`

### PR DESCRIPTION
**Pull request by lexmag**
Follow-up on #3382.

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/capture_io.ex [hunk 1]
```diff
@@ -124,9 +124,18 @@ defmodule ExUnit.CaptureIO do
     {:ok, capture_io} = StringIO.open(input)
     Process.register(capture_io, device)
 
+    flag = Process.flag(:trap_exit, true)
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: We don't want to force a trap exit because this is surprising behaviour to a user :(

Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: Totally agree, it's just a temporary thing to make the test pass. :smile:

**Comment by josevalim** [comment]
I think we can tackle this one the same way we have done in #3445? I.e. let the client do most of the work, the server will only clean up in case something goes wrong?

**Comment by lexmag** [comment]
Not sure if there is anything more what we can keep on client. I would like your feedback on it.

**Comment by fishcakez** [comment]
This can be simplified a bit.  The `IEx.Server` is only there to provide the lock and to cleanup on `:DOWN`. The `IEx.Server` can do the lookup of the previous device (to prevent race conditions) and monitor the `StringIO` device before replying with the previous device and reference. Then if it gets a `:DOWN` it just needs to try to register the previous device, if this raises an `ArgumentError` it can ignore it because either the previous device is dead or the name is registered by another process (possibly the previous device but definitely not the `StringIO`).

**Comment by fishcakez** [comment]
Bad choice of phrase, I meant the `IEx.Server` can be simplified by moving more to the client.

**Comment by lexmag** [comment]
@fishcakez thanks for feedback, and you meant `ExUnit.Server` of course. :smile:

**Comment by fishcakez** [comment]
Yeah :)

**Comment by lexmag** [comment]
I've just tried to move some bits to the client side (see c855d31), don't seem it improves implementation: `Process.unregister` in two places (we still need it in `:DOWN` for an exit signal case).

File: lib/ex_unit/lib/ex_unit/capture_io.ex [hunk 2]
```diff
@@ -110,30 +110,29 @@ defmodule ExUnit.CaptureIO do
...
+        raise "could not find IO device registered at #{inspect device}"
+      {:error, :already_captured} ->
+        _ = StringIO.close(string_io)
+        raise "IO device registered at #{inspect device} is already captured"
+    end
 
     try do
-      do_capture_io(capture_io, fun)
+      Process.unregister(device)
+      Process.register(pid, device)
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: Should this be `string_io`, instead of `pid`?

File: lib/ex_unit/lib/ex_unit/server.ex [hunk 3]
```diff
@@ -116,11 +142,29 @@ defmodule ExUnit.Server do
...
+    case Map.pop(refs, ref) do
+      {{device, pid}, refs} ->
+        devices = HashSet.delete(devices, device)
+        Process.demonitor(ref, [:flush])
+        try do
+          Process.unregister(device)
+        rescue
+          ArgumentError -> :ok
+        after
+          Process.register(pid, device)
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/server.ex]: If register raises here another process has registered as `device`. It is not the `string_io` process so we can rescue `ArgumentError` and ignore it.

File: lib/ex_unit/lib/ex_unit/capture_io.ex [hunk 4]
```diff
@@ -110,30 +110,29 @@ defmodule ExUnit.CaptureIO do
...
+      do_capture_io(string_io, fun)
     after
       try do
         Process.unregister(device)
       rescue
         ArgumentError -> nil
       end
-      Process.register(original_io, device)
```
Comment (fishcakez) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: We can try to register `pid` here?

File: lib/ex_unit/lib/ex_unit/capture_io.ex (hunk 2 above)
Comment (lexmag) [inline lib/ex_unit/lib/ex_unit/capture_io.ex]: Definitely it should, it's copied verbatim here to see how it looks like on client.

**Comment by josevalim** [comment]
This one looks good. Please rebase so we can merge!

**Comment by lexmag** [comment]
@josevalim this is ready for merge.

**Pull request merged**